### PR TITLE
GraphQL - Implement session api

### PR DESCRIPTION
**Pull request by Premwoik**
This PR addresses [MIM-1581](https://erlangsolutions.atlassian.net/browse/MIM-1581) and adds support for the session category.

All session command functions have been extracted to the `mongoose_session_api` module.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/103608) / small_tests / e274c2281c5df6c0b39a6ba1f9222b38b2de5747
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/103608/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/103608/small_tests.23.1/small/ct_run.mongooseim@localhost.2022-01-28_17.13.25/lib.mongooseim.logs/run.2022-01-28_17.13.26/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/103957) / small_tests / a16e25e5c5620dad30f9416b402dd4086737dcd4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/103957/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/103957/small_tests.23.1/small/ct_run.mongooseim@localhost.2022-01-31_15.35.29/lib.mongooseim.logs/run.2022-01-31_15.35.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104070) / small_tests / 4a2d87b314f3fd9463d5e30095ea81f41a991bb4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104070/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/104070/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-31_20.20.50/lib.mongooseim.logs/run.2022-01-31_20.20.50/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104119) / small_tests / dc659047f0de89a2bdf6a9eae59117040229d9bc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104119/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/104119/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-01_09.21.24/lib.mongooseim.logs/run.2022-02-01_09.21.24/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104368) / small_tests / 49e72cdd5a341846bd8dda867c61ec1accd756f4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104368/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/104368/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-01_16.55.18/lib.mongooseim.logs/run.2022-02-01_16.55.18/suite.log.html)

**Comment by mongoose-im** [comment]
[internal_mnesia_24](https://circleci.com/gh/esl/MongooseIM/104390) / internal_mnesia / 97e7244d83f7fb404f1564073149ed1fb3a0eb4b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104390/internal_mnesia.24.0.5)

File: priv/graphql/schemas/admin/admin_schema.gql [hunk 1]
```diff
@@ -26,5 +28,7 @@ type AdminMutation @protected{
   account: AccountAdminMutation
   "Modify domains properties or create/destroy domain"
   domains: DomainAdminMutation
+  "Session management"
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/admin_schema.gql:31]: It would be great to use either noun or verb form consistently, i.e. either "Manage sessions", "Modify domains" or "Account management", "Domain management", same for queries. Maybe the noun form would be better for categories?

File: priv/graphql/schemas/admin/session.gql [hunk 2]
```diff
@@ -0,0 +1,81 @@
+"""
+Allow admin to get information about sessions.
+"""
+type SessionAdminQuery @protected{
+  "Get the list of established sessions for a specified domain or global"
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/session.gql:5]:
```suggestion
  "Get the list of established sessions for a specified domain or globally"
```
I think it's grammatically incorrect - an adverb is needed here. Same would apply below.

File: priv/graphql/schemas/admin/session.gql [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  "Get the number of established sessions for a specified domain or global"
+  countSessions(domain: String): Int 
+  "Get information about all sessions of a user"
+  listUserSessions(user: JID!): [Session!]
+  "Get the number of resources of a user"
+  countUserResources(user: JID!): Int
+  "Get the resource string of a n-th session number"
+  getUserResource(user: JID!, number: Int): String
+  "Get the list of logged users with this status for a specified domain or global"
+  listStatusUsers(domain: String, status: String!): [StatusUser!]
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/session.gql:16]: Could you reword this to `listUsersWithStatus`? `StatusUsers` sounds awkward, like they were using some status.

File: priv/graphql/schemas/admin/session.gql [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+type SessionAdminQuery @protected{
+  "Get the list of established sessions for a specified domain or global"
+  listSessions(domain: String): [Session!]
+  "Get the number of established sessions for a specified domain or global"
+  countSessions(domain: String): Int 
+  "Get information about all sessions of a user"
+  listUserSessions(user: JID!): [Session!]
+  "Get the number of resources of a user"
+  countUserResources(user: JID!): Int
+  "Get the resource string of a n-th session number"
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/session.gql:13]:
```suggestion
  "Get the resource string of the n-th session of a user"
```

File: priv/graphql/schemas/admin/session.gql [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+"Modify session payload"
+type SessionPayload{
+  "Full JID with username, server and resource"
+  jid: JID!
+  "Result message"
+  message: String!
+}
+
+"Short info about user's status"
+type StatusUser{
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/session.gql:74]:
```suggestion
type UserStatus{
```

File: src/graphql/admin/mongoose_graphql_session_admin_mutation.erl [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    Status2 = null_to_empty(Status),
+    Priority2 = null_to_empty(Priority),
+    Result = mongoose_session_api:set_presence(JID, Type, Show2, Status2, Priority2),
+    format_session_payload(Result, JID).
+
+%% Internal
+
+-spec null_to_empty(null | integer() | binary()) -> binary().
+null_to_empty(null) -> <<>>;
+null_to_empty(Int) when is_integer(Int)-> integer_to_binary(Int);
```
Comment (chrzaszcz) [inline src/graphql/admin/mongoose_graphql_session_admin_mutation.erl:35]:
```suggestion
null_to_empty(Int) when is_integer(Int) -> integer_to_binary(Int);
```

File: src/mongoose_session_api.erl [hunk 7]
```diff
@@ -1,12 +1,179 @@
...
+                        Prio    :: ejabberd_sm:priority(),
+                        Status  :: status()}.
+
+-type session_info() :: {USR :: binary(),
+                                Conn :: binary(),
+                                IPS :: binary(),
+                                Port :: inet:port_number(),
+                                Prio :: ejabberd_sm:priority(),
+                                NodeS :: binary(),
+                                Uptime :: integer()}.
```
Comment (chrzaszcz) [inline src/mongoose_session_api.erl:50]: Please fix indentation

File: src/mongoose_session_api.erl [hunk 8]
```diff
@@ -1,12 +1,179 @@
...
+set_presence(JID, Type, Show, Status, Priority) ->
+    Pid = ejabberd_sm:get_session_pid(JID),
+    USR = jid:to_binary(JID),
+    US = jid:to_binary(jid:to_bare(JID)),
+
+    Children = maybe_pres_status(Status,
+                                 maybe_pres_priority(Priority,
+                                                     maybe_pres_show(Show, []))),
+    Message = {xmlstreamelement,
+               #xmlel{ name = <<"presence">>,
```
Comment (chrzaszcz) [inline src/mongoose_session_api.erl:170]:
```suggestion
               #xmlel{name = <<"presence">>,
```

File: src/mongoose_session_api.erl [hunk 9]
```diff
@@ -33,3 +204,54 @@ prepare_reason(Reason) when is_list(Reason) ->
...
     Reason.
+
+%% Internal
+
+-spec get_status_list([session()], status()) -> [status_user_info()].
+get_status_list(Sessions0, StatusRequired) ->
+    Sessions = [ {catch ejabberd_c2s:get_presence(Pid), S, P}
+                 || #session{sid = {_, Pid}, usr = {_, S, _}, priority = P}
+                    <- Sessions0 ],
+
```
Comment (chrzaszcz) [inline src/mongoose_session_api.erl:215]:
```suggestion
                 || #session{sid = {_, Pid}, usr = {_, S, _}, priority = P} <- Sessions0 ],
```

**Review by chrzaszcz — APPROVED** [review]
The code looks good to me. I know it was not in the doc, but I think there should be a command allowing the user to set their own presence - just like the admin command. Otherwise the API is kind of incomplete.

**Review by chrzaszcz — COMMENTED** [review]
I wanted to comment, but I clicked 'approve' by mistake, apparently the code is just good 🙂

**Comment by mongoose-im** [comment]
[mssql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/104650) / odbc_mssql_mnesia / d34f161cbd965fb3346009d6bacc798d8fe19567
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104650/odbc_mssql_mnesia.24.0.5)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/104782) / small_tests / 44126219fc17f06de786281ce6c01e610d8352d0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3521/104782/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3521/104782/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-02-03_10.16.51/lib.mongooseim.logs/run.2022-02-03_10.16.51/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
Looks good!
